homeassistant/components/google/calendar.py

```diff
@@ -265,48 +338,44 @@ def _event_filter(self, event: Event) -> bool:
             return True
         return event.transparency == OPAQUE
 
+    async def async_added_to_hass(self) -> None:
+        """When entity is added to hass."""
+        await super().async_added_to_hass()
+        self._apply_coordinator_update()
+
     async def async_get_events(
         self, hass: HomeAssistant, start_date: datetime, end_date: datetime
     ) -> list[CalendarEvent]:
         """Get all events in a specific time frame."""
-
-        request = ListEventsRequest(
-            calendar_id=self.calendar_id,
-            start_time=start_date,
-            end_time=end_date,
-            search=self._search,
-        )
-        result_items = []
-        try:
-            result = await self.calendar_service.async_list_events(request)
-            async for result_page in result:
-                result_items.extend(result_page.items)
-        except ApiException as err:
-            raise HomeAssistantError(str(err)) from err
+        result_items = await self.coordinator.async_get_events(start_date, end_date)
         return [
             _get_calendar_event(event)
             for event in filter(self._event_filter, result_items)
         ]
 
-    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def _apply_coordinator_update(self) -> None:
+        """Copy state from the coordinator to this entity."""
+        events = self.coordinator.data
+        self._event = _get_calendar_event(next(iter(events))) if events else None
+        if self._event:
+            (self._event.summary, self._offset_value) = extract_offset(
+                self._event.summary, self._offset
+            )
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._apply_coordinator_update()
+        super()._handle_coordinator_update()
+
     async def async_update(self) -> None:
-        """Get the latest data."""
-        request = ListEventsRequest(calendar_id=self.calendar_id, search=self._search)
-        try:
-            result = await self.calendar_service.async_list_events(request)
-        except ApiException as err:
-            _LOGGER.error("Unable to connect to Google: %s", err)
-            return
-
-        # Pick the first visible event and apply offset calculations.
-        valid_items = filter(self._event_filter, result.items)
-        event = copy.deepcopy(next(valid_items, None))
-        if event:
-            (event.summary, offset) = extract_offset(event.summary, self._offset)
-            self._event = _get_calendar_event(event)
-            self._offset_value = offset
-        else:
-            self._event = None
+        """Disable update behavior.
+
+        This relies on the coordinator callback update to write home assistant
```

Comment:
If you know the calendar to be out of sync, wouldn't be helpful to fetch the latest events from the server by calling the update entity service?

Reply:
I guess we can't distinguish here if it's from polling or a service call.

Reply:
I think the fundamental problem is the calendar entity model is incorrect, using the same entity for both a calendar and matching events on the calendar. I've proposed an architecture discussions about calendar events.

Reply:
(It's a fairly large breaking change to split this up, however, so not sure how feasible it would be.)

If we want to add extra retries in the data coordinator, should we just do that as a feature of the data update coordinator? I'm not sure calendars are necessarily special in that regard.  If the data is out of sync, then it's going to be marked as unavailable, like other entities, which seems nice and simple. 

Part of my plan here is to replace this coordinator with an alternative one that will sync deltas from the server.  It can then serve events out of the local calendar store.  When that is introduced, i'm not exactly sure the best way to handle unavailability (e.g. maybe only if it fails a certain # of times?).

Reply:
Perhaps the proposal is too long that my actual proposal is missing but it's at the end: 'I lean towards adding a "Calendar Event" helper': https://github.com/home-assistant/architecture/discussions/776

Related issue: https://github.com/home-assistant/core/issues/44145